var/spack/repos/builtin/packages/fsl/package.py

```diff
@@ -8,49 +8,163 @@
 import os
 
 
-class Fsl(Package):
+class Fsl(Package, CudaPackage):
     """FSL is a comprehensive library of analysis tools for FMRI, MRI and DTI
-       brain imaging data.
+       brain imaging data."""
 
-       Note: A manual download is required for FSL.
-       Spack will search your current directory for the download file.
-       Alternatively, add this file to a mirror so that Spack can find it.
-       For instructions on how to set up a mirror, see
-       http://spack.readthedocs.io/en/latest/mirrors.html"""
+    # NOTE: A manual download is required for FSL.  Spack will search your
+    # current directory for the download file.  Alternatively, add this file to
+    # a mirror so that Spack can find it.  For instructions on how to set up a
+    # mirror, see http://spack.readthedocs.io/en/latest/mirrors.html
 
     homepage = "https://fsl.fmrib.ox.ac.uk"
     url      = "file://{0}/fsl-5.0.10-sources.tar.gz".format(os.getcwd())
     manual_download = True
 
-    version('5.0.10', '64823172a08aad679833240ba64c8e30')
+    version('6.0.4', sha256='58b88f38e080b05d70724d57342f58e1baf56e2bd3b98506a72b4446cad5033e')
+    version('5.0.10', sha256='ca183e489320de0e502a7ba63230a7f55098917a519e8c738b005d526e700842')
 
     depends_on('python', type=('build', 'run'))
     depends_on('expat')
     depends_on('libx11')
     depends_on('glu')
-    depends_on('zlib')
-    depends_on('libpng')
-    depends_on('boost')
-    depends_on('sqlite')
-
-    conflicts('%gcc@6:', when='@5.0.10')
+    depends_on('libiconv')
+    depends_on('openblas', when='@6:')
+    depends_on('vtk@:8')
+
+    conflicts('%gcc@:4.7,5.5:')
+    conflicts('^cuda@9.2:')
+    conflicts('cuda_arch=none', when='+cuda',
+              msg='must select a CUDA architecture')
+    conflicts('platform=darwin',
+              msg='currently only packaged for linux')
+
+    patch('build_log.patch')
+    patch('eddy_Makefile.patch', when='@6:')
+    patch('fslpython_install_v5.patch', when='@:5')
+    patch('fslpython_install_v6.patch', when='@6:')
+    patch('fsl_sub_v5.patch', when='@:5')
+    patch('fsl_sub_v6.patch', when='@6:')
+    patch('iconv.patch')
```

Comment:
Can you add a comment saying where the patches are coming from?

Reply:
I created the patches. I am not sure how you want that indicated.

Reply:
Have these patches been contributed upstream?

Reply:
I normally do contribute upstream. However, the distribution of FSL requires registration and so it is not clear to me where patches would be submitted. Also, most of these patches would not apply upstream anyway.

The following patches are specific to spack:
- build_log.patch
- fslpython_install_v6.patch

The following is only necessary for the older version so is no longer relevant upstream.
- fslpython_install_v5.patch

The following are special cases. I referenced this in my initial commit but I will add a comment to the package recipe for this to be clearer.
- fsl_sub_v5.patch
- fsl_sub_v6.patch
These patches disable FSL's attempts to try to submit a subset of FSL computations to an SGE queue system. That mechanism only works for SGE and requires someone to edit the `fsl_sub` script to accommodate their system. These patches disable that and allow the `fsl_sub` script to behave the same on all systems, and without further modification, whether the computation is submitted to a "local" system, like a workstation, or as a batch job to a cluster queueing system, regardless of queue system type. These patches make no attempt to "fix" the behavior to make it work on other queueing systems, etc., and thus would be of no interest upstream.

If there is disagreement to having the fsl_sub patches as I have done then perhaps a variant is an option. I will also add that FSL needs a conda environment to be installed after installation, which is kind of ugly. Because of that, although we use FSL at our site, I probably would not have submitted this as a PR if there was not an FSL package already in spack.